synapse/handlers/e2e_keys.py

```diff
@@ -155,16 +160,25 @@ async def query_devices(self, query_body, timeout, from_user_id):
                             unsigned["device_display_name"] = device_display_name
                         user_devices[device_id] = result
 
+            # fetch missing cross-signing keys from federation.
+            for user_id in remote_queries.keys():
+                has_cross_master = user_id in cross_signing_keys["master_keys"]
+                has_cross_selfsigning = (
+                    user_id in cross_signing_keys["self_signing_keys"]
+                )
+
+                # check if we do have a master/self-signing key, but not the other.
+                # for each user we want the master _and_ the self-signing key,
+                # so we fetch those keys from federation
+                if has_cross_master ^ has_cross_selfsigning:
```

Comment:
why do we not do this if both `has_cross_master` and `has_cross_selfsigning` are `False`?

Reply:
now we do it in this case, too.